efaar_benchmarking/data_loading.py

```diff
@@ -1,17 +1,98 @@
 
 import numpy as np
+import pandas as pd
 import scanpy as sc
 import wget
+from sklearn.covariance import EllipticEnvelope
+
+CP_FEATURE_FORMATTER = (
+    "s3://cellpainting-gallery/cpg0016-jump/"
+    "{Metadata_Source}/workspace/profiles/"
+    "{Metadata_Batch}/{Metadata_Plate}/{Metadata_Plate}.parquet"
+)
+
+
+def load_cpg16_crispr(data_path="data/", filter_by_intensity=True, filter_by_cell_count=True, drop_image_cols=True):
+    plate_file_name = "plate.csv.gz"
+    well_file_name = "well.csv.gz"
+    crispr_file_name = "crispr.csv.gz"
+    plate_file_path = os.path.join(data_path, plate_file_name)
+    well_file_path = os.path.join(data_path, well_file_name)
+    crispr_file_path = os.path.join(data_path, crispr_file_name)
+    if not (os.path.exists(plate_file_path) and os.path.exists(well_file_path) and os.path.exists(crispr_file_path)):
+        source_path = "https://zenodo.org/records/7661296/files/jump-cellpainting/datasets-v0.5.0.zip?download=1"
+        path_to_zip_file = data_path + "tmp.zip"
+        wget.download(source_path, path_to_zip_file)
+        with zipfile.ZipFile(path_to_zip_file, "r") as zip_ref:
+            for name in zip_ref.namelist():
+                if name.endswith(plate_file_name) or name.endswith(well_file_name) or name.endswith(crispr_file_name):
+                    with tempfile.TemporaryDirectory() as temp_dir:
+                        zip_ref.extract(name, temp_dir)
+                        shutil.move(os.path.join(temp_dir, name), os.path.join(data_path, os.path.basename(name)))
+        os.remove(path_to_zip_file)
+
+    plates = pd.read_csv(plate_file_path)
+    crispr_plates = plates.query("Metadata_PlateType=='CRISPR'")
+    wells = pd.read_csv(well_file_path)
+    well_plate = wells.merge(crispr_plates, on=["Metadata_Source", "Metadata_Plate"])
+    crispr = pd.read_csv(crispr_file_path)
+    metadata = well_plate.merge(crispr, on="Metadata_JCP2022")
+
+    features_file_path = os.path.join(data_path, "cpg_features.parquet")
+    if not os.path.exists(features_file_path):
+
+        def load_plate_features(path: str):
+            return pd.read_parquet(path, storage_options={"anon": True})
+
+        cripsr_plates = metadata[
+            ["Metadata_Source", "Metadata_Batch", "Metadata_Plate", "Metadata_PlateType"]
+        ].drop_duplicates()
+        features = []
+        with ThreadPoolExecutor(max_workers=10) as executer:
+            future_to_plate = {
+                executer.submit(
+                    load_plate_features, CP_FEATURE_FORMATTER.format(**row.to_dict())
+                ): CP_FEATURE_FORMATTER.format(**row.to_dict())
+                for _, row in cripsr_plates.iterrows()
+            }
+            for future in as_completed(future_to_plate):
+                features.append(future.result())
+        pd.concat(features).to_parquet(features_file_path)
+    features = pd.read_parquet(features_file_path).dropna(axis=1)
+    if filter_by_intensity:
```

Comment:
I think it might be worth splitting out the loading from the processing

Reply:
Done in the new commit.